iModelCore/ECDb/ECDb/DbMappingManager.cpp

```diff
@@ -698,47 +698,83 @@ BentleyStatus DbMappingManager::Classes::MapUserDefinedIndexes(SchemaImportConte
-        return SUCCESS;
+    const auto checkDerivedClassIndexes = ProfileVersion(versionStr.c_str()) <= BeVersion(4, 0, 0, 2);
 
-    if (SUCCESS != ValidateUserDefinedIndexes(importCtx, classMap))
-        return ERROR;
+    for (auto& indexCA : indexCAs)
+        {
+        auto indexName = indexCA.GetName();
+        auto isSystemIndex = systemIndexes.find(indexName) != systemIndexes.end();
 
-    if (SUCCESS != importCtx.GetSchemaManager().GetDbSchema().LoadIndexDefs())
-        return ERROR;
+        // Special case: Older imodels ( <= 4.0.0.2) do not have a common unique index on the base TPH table to handle all the derived classes at once.
+        // In such a case, check if a unique index exists without the classid and select that as the index to update instead.
+        if (checkDerivedClassIndexes && !isSystemIndex && indexName.EndsWithI("_sourcetargetclassid"))
+            {
+            auto oldIndexName = indexName.substr(0, indexName.size() - Utf8String("classid").size());
+            if (isSystemIndex = systemIndexes.find(oldIndexName) != systemIndexes.end(); isSystemIndex)
+                indexCA.SetIndexName(oldIndexName);
+            }
 
-    auto& table = classMap.GetJoinedOrPrimaryTable();
-    for (DbIndexListCustomAttribute::DbIndex const& indexCA : indexCAs)
-        {
-        table.RemoveIndexDef(indexCA.GetName());
+        if (isSystemIndex)
+            {
+            if (!classMap.GetClass().IsRelationshipClass() || !classMap.GetMapStrategy().GetTphInfo().IsValid())
+                {
+                importCtx.Issues().ReportV(IssueSeverity::Error, IssueCategory::BusinessProperties, IssueType::ECDbIssue, ECDbIssueId::ECDb_0680,
+                    "The system index '%s' specified in the custom attribute 'DbIndexList' for class '%s' is not eligible for a modification. System-generated indexes can only be modified in relationship classes with TPH map strategy.",
+                    indexCA.GetName().c_str(), classMap.GetClass().GetFullName());
+                return ERROR;
+                }
+            }
+        if (updateIndexes || isSystemIndex)
+            table.RemoveIndexDef(indexCA.GetName());
 
-        if (SUCCESS != MapUserDefinedIndex(importCtx, classMap, indexCA))
+        if (SUCCESS != MapIndex(importCtx, classMap, indexCA, isSystemIndex))
             return ERROR;
+
+        // Special case: Since older imodels ( <= 4.0.0.2) have individual unique indexes for the derived classes as well, just updating the one for the base class is not enough.
+        // As we have already updated the unique index in the base class to handle the uniqueness in the derived classes as well, we can now remove the unique indexes in the derived classes.
+        if (checkDerivedClassIndexes && isSystemIndex && indexCA.IsUnique())
+            {
+            if (const auto allDerivedClasses = importCtx.GetECDb().Schemas().GetAllDerivedClassesInternal(classMap.GetClass()); allDerivedClasses.IsValid())

```

Comment:
Nested IFs make code less readable. You can use a short circuit pattern where you `continue`, `break` or `return` when a rule is false. This avoids nesting IFs and thus intending code.

Instead of following

```c
while() {
  if (a) {
    if (b) {
     // work
    }
  }
}
```
use
```c
while () {
  if (!a) continue;
  if (!b) continue;
  // work
}
```